server/src/main/java/org/elasticsearch/common/util/CollectionUtils.java

```diff
@@ -100,44 +99,68 @@ public static int[] toArray(Collection<Integer> ints) {
      * @param messageHint A string to be included in the exception message if the call fails, to provide
      *                    more context to the handler of the exception
      */
-    public static void ensureNoSelfReferences(Object value, String messageHint) {
-        Iterable<?> it = convert(value);
-        if (it != null) {
-            ensureNoSelfReferences(it, value, Collections.newSetFromMap(new IdentityHashMap<>()), messageHint);
+    public static void ensureNoSelfReferences(final Object value, final String messageHint) {
+        ensureNoSelfReferences(value, Collections.newSetFromMap(new IdentityHashMap<>()), messageHint);
+    }
+
+    private static void ensureNoSelfReferences(final Object value, final Set<Object> ancestors, final String messageHint) {
+        if (value == null || value instanceof String || value instanceof Number || value instanceof Boolean) {
+            // noop
+        } else if (value instanceof Map<?, ?> m && m.isEmpty() == false) {
+            ensureNoSelfReferences(m, ancestors, messageHint);
+        } else if ((value instanceof List<?> l) && l.isEmpty() == false) {
+            ensureNoSelfReferences(l, ancestors, messageHint);
+        } else if ((value instanceof Set<?> s) && s.isEmpty() == false) {
+            ensureNoSelfReferences(s, ancestors, messageHint);
+        } else if ((value instanceof Iterable<?> i) && (value instanceof Path == false)) {
```

Comment:
Could we combine list/set/iterable here? As this is a general utility method, I imagine there are enough implementations of list/set/iterable being used here by themselves that it's using the many-implementations vtable lookup anyway. I may be wrong though.

Reply:
Probably, I might have gotten a little overexcited. 😉

I'll write a combined version and run it through the benchmarks to see if it costs us. If not, then simpler is better for sure.

Reply:
Indeed, no obvious change in performance for my workload without those specialized implementations, I've removed them in 664e29f0da0ea5f5e4a621cec3298f3c64094f69.